esp/services/ws_workunits/ws_workunitsHelpers.cpp

```diff
@@ -1049,25 +1049,81 @@ void WsWuInfo::getInfo(IEspECLWorkunit &info, unsigned long flags)
     getWorkflow(info, flags);
 }
 
+StringBuffer& WsWuInfo::getWorkunitProcessLogPath(const char* process, StringBuffer& path)
+{
+    Owned<IPropertyTreeIterator> procs = cw->getProcesses(process, nullptr);
+    if (!procs->first())
+        return path;
+
+    StringBuffer logSpec;
+    procs->query().getProp("@log", logSpec);
+    if (!logSpec.length())
+        return path;
+
+    splitFilename(logSpec, nullptr, &path, nullptr, nullptr);
+    return path;
+}
+
+unsigned WsWuInfo::getNumberOfSlavesByClusterName()
+{
+    StringAttr clusterName(cw->queryClusterName());
+    if (!clusterName.length()) //Cluster name may not be set yet
+        return 0;
+
+    Owned<IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(clusterName.str());
+    if (!clusterInfo)
+    {
+        IWARNLOG("Cannot find TargetClusterInfo for workunit %s", cw->queryWuid());
+        return 0;
+    }
+
+    return clusterInfo->getNumberOfSlaveLogs();
+}
+
+unsigned WsWuInfo::getNumberOfSlavesFromZAPFiles()
+{
+    StringBuffer zapWUFolderPath;
+    getWorkunitProcessLogPath("Thor", zapWUFolderPath);
+
+    unsigned numberOfSlaves = 0;
+    Owned<IFile> zapFileFolder = createIFile(zapWUFolderPath);
+    Owned<IDirectoryIterator> thorSlaveFiles = zapFileFolder->directoryFiles("*_thorslave.*");
+    ForEach(*thorSlaveFiles)
+    {
+        const char* pStr = strstr(thorSlaveFiles->query().queryFilename(), "_thorslave.");
+        pStr += 11;
+        const char* slaveNumStrEnd = strchr(pStr, '.');
+        if (!slaveNumStrEnd)
+            continue;
+
+        StringBuffer slaveNumStr;
+        slaveNumStr.append(slaveNumStrEnd - pStr, pStr);
+        unsigned slaveNum = atoi(slaveNumStr);
+        if (slaveNum > numberOfSlaves)
+            numberOfSlaves = slaveNum;
+    }
+    return numberOfSlaves;
+}
+
 unsigned WsWuInfo::getWorkunitThorLogInfo(IArrayOf<IEspECLHelpFile>& helpers, IEspECLWorkunit &info, unsigned long flags, unsigned& helpersCount)
 {
     unsigned countThorLog = 0;
 
     IArrayOf<IConstThorLogInfo> thorLogList;
     if (cw->getWuidVersion() > 0)
     {
-        StringAttr clusterName(cw->queryClusterName());
-        if (!clusterName.length()) //Cluster name may not be set yet
-            return countThorLog;
-
-        Owned<IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(clusterName.str());
-        if (!clusterInfo)
+        unsigned numberOfSlaveLogs = 0;
+        StringAttr thorMasterLogDateSearchString;
+        if (debugImportedStr.length()) //WU imported from a ZAP report
         {
-            IWARNLOG("Cannot find TargetClusterInfo for workunit %s", cw->queryWuid());
-            return countThorLog;
+            numberOfSlaveLogs = getNumberOfSlavesFromZAPFiles();
+            thorMasterLogDateSearchString.set("_thormaster.");
+        }
+        else
+        {
+            numberOfSlaveLogs = getNumberOfSlavesByClusterName();
+            thorMasterLogDateSearchString.set("/thormaster.");
```

Comment:
can u always deduce # of slaves from the log (names) if number of slaves is required?
What if slave logs aren't available, because zap report didn't contain them?

What is numberOfSlaveLogs needed for?
I see the:
```
                thorLog->setNumberSlaves(numberOfSlaveLogs);
```
..below. Is that so Eclwatch can go to a particular slave log?

If so, could deduce # of slaves from slave log names if available in imported helpers.
Just a thought, perhaps workunit should always, contain group and pattern to slave logs, so that doesn't have to go lookup stuff in environment which may be out of date anyway.
Then an imported workunit would not be particularly different, would also contain group and pattern in same way.

Reply:
There's:
```
1142:            StringBuffer groupName;
1143:            getClusterThorGroupName(groupName, processName.str());
```

that will likely not make much sense for an imported version.
i.e. the processName will not exist in the local environment.

what is groupName needed for?

Reply:
> .below. Is that so Eclwatch can go to a particular slave log?
Yes, ECLWatch has to display a number which a user may choose to download a slave log.  
> what is groupName needed for?
The groupName is used for standard (not imported) WU to get slave log.

Reply:
So what if instead there were methods in the workunit like:
```
IGroup *IConstWorkUnit::getThorGroup();
```
and:
```
const char *IConstWorkUnit::getSlaveLogPattern();
```

and Thor could published the group and the slave log pattern to the workunit.
For legacy workunits (that didn't have published group/pattern), it could as you do, get the group based on getClusterThorGroupName, and the slave pattern from the environment directories.

Then this code could be agnostic to where workunit came from

Reply:
Here, I need: ThorGroup, # of slaves, and the dates of the thor logs. The getThorGroup() returns the ThorGroup. What will getSlaveLogPattern() return?

Reply:
If we have the logPerWorkunit, I still need to know the number of slaves (for example, 50) so that ECLwatch can display a list of numbers (from 1 to 50) for a user to download a specific slave log. So, Should the pattern be: /var/log/HPCCSystems/mythor/thorslave.*_of_50.2019_11_1[2-3].log or /var/log/HPCCSystems/mythor/W20191113-140205/thorslave.*_of_50.log ?